exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/common/HashTableTemplate.java

```diff
@@ -111,6 +111,31 @@
 
   private int resizingTime = 0;
 
+
+  // Given hashtable index, figure out the corresponding index in the batch.
+  public int idxInBatch(int idx) {
```

Comment:
This is complexity comes from the heterogenous BatchHolder sizes. Additionally there is a performance penalty for calculating indexes as the HashTable gets larger. As my comment above mentions, I don't think there is any benefit to having different sizes for batch holders. We should compute a reasonable size based on the first set of input data, and then use the same BatchHolder size everywhere.

Reply:
The original code referred to location as a pair: < batch-num ; index-in-batch >
This was condensed into a 4 byte "int", each element in a 2 byte.
With this model, any batch can be of a different size, and access be done as before (e.g. **& BATCH_MASK**). Only "did we reach the limit" checks need to be done against the per-batch limit.
The code that accesses the hash table is super hot code (per each row inserted, searched, etc). Thus running thru such loops is a big performance penalty !!!!!